### ISSUE TITLE
There seems to be randomness in GPU. How can I control it?

### ISSUE THREAD

**Issue by jellyheadandrew** [PaddlePaddle/Paddle#49125]
### 请提出你的问题 Please ask your question

Hello, I was trying to **control randomness 100% to reproduce "exactly same results" that shows on training log** when training, similar to what I have asked earlier: [https://github.com/PaddlePaddle/Paddle/issues/49058](https://github.com/PaddlePaddle/Paddle/issues/49058).

_Please refer to pytorch version of this problem: [https://github.com/pytorch/pytorch/issues/7068](https://github.com/pytorch/pytorch/issues/7068)_

I tried 
-seeded random `random.seed(args.seed)`
-seeded np.random `np.random.seed(args.seed)`
-seeded paddle `paddle.seed(args.seed)`

All of the "augmentation" (such as random crop) or paddle operations resulted the same, but I found randomness in GPU. For instance, in "Segformer" code, I tried printing out the value of features as below. **(note that I do not use "dropout" in the model)**

![image](https://user-images.githubusercontent.com/80747187/208110869-76f460bd-1e72-4f97-bc95-cfd92f3cff00.png)

However, when I ran the code twice, the printed out results were very similar but slightly different.

First run
![image](https://user-images.githubusercontent.com/80747187/208111602-b1682fbc-792c-4bb0-bd4e-054c5e51a403.png)

Second run
![image](https://user-images.githubusercontent.com/80747187/208111629-f47c9498-25f2-4c62-b6ef-19f44baacc4c.png)

I think the **GPU CUDA's randomness/error** causes this problem, and this randomness accumulates and returns different training results. **Am I right? Are there any ways to control this randomness/error from GPU?**

**Also, even if I do use the dropout or any randomness that appears in any optimizer, model, etc., I want to fully control randomness as in [https://github.com/pytorch/pytorch/issues/7068](https://github.com/pytorch/pytorch/issues/7068). Are there any ways?**

_**Update: I have added below codes also, but still doesn't reproduce same results when training**_

    random.seed(args.seed)
    np.random.seed(args.seed)
    paddle.seed(args.seed)
    os.environ['FLAGS_cudnn_deterministic']="1"
    import paddle.fluid as fluid
    fluid.default_startup_program().random_seed = args.seed
    fluid.default_main_program().random_seed=args.seed

**Comment by jellyheadandrew** [comment]
Also, even if I do use the dropout or any randomness that appears in any optimizer, model, etc., I want to fully control randomness as in [https://github.com/pytorch/pytorch/issues/7068](https://github.com/pytorch/pytorch/issues/7068).

**Comment by jellyheadandrew** [comment]
I tried

`paddle.seed(args.seed)`
`np.random.seed(args.seed)`
`random.seed(args.seed`
`paddle.set_flags({'FLAGS_cudnn_deterministic: True, FLAGS_cpu_determinitstic: True})`
`fluid.default_startup_program().random_seed = args.seed`
`fluid.default_main_program().random_seed = args.seed`
`os.environ['PYTHONHASHSEED']=str(args.seed)`

but still doesn't work. I am in urgent need :sob: Plz help!!

**Comment by w5688414** [comment]
> I tried
> 
> `paddle.seed(args.seed)` `np.random.seed(args.seed)` `random.seed(args.seed` `paddle.set_flags({'FLAGS_cudnn_deterministic: True, FLAGS_cpu_determinitstic: True})` `fluid.default_startup_program().random_seed = args.seed` `fluid.default_main_program().random_seed = args.seed` `os.environ['PYTHONHASHSEED']=str(args.seed)`
> 
> but still doesn't work. I am in urgent need 😭 Plz help!!

Hi, @jellyheadandrew , can you try to run the program on cpu instead of gpu, since it's not easy to reproduce the results on gpu right now

**Comment by w5688414** [comment]
Hi, @jellyheadandrew, Can you provide sample code for reproducing the results, we want to analysis your code in depth.
